File: univ/activeMeasurement/AuxiliaryCode/数据结构.py
from typing import List
from typing import Mapping
from typing import Tuple

class TimeStamp:
    second: int

class Switch:
    id: int
    
class RecordPerSwitch:
    IPid: int
    time_stamp: TimeStamp
    
class PathRecord:
    # 长度和对应的Flow.path相同
    record: List[RecordPerSwitch]

class Flow:
    id: int
    # 标识每一跳
    path: List[Switch]
    # 标识每次记录
    record: List[PathRecord]

class RecordOnSwitch:
    # Map: flow_id->( Map: ipid-> time_stamp或标识不存在 )
    Mapping[int, Mapping[int, TimeStamp]]

class Link:
    # 也许可以换成两端的名字？反正是个好用的码
    id: int
    record_of_input_node: RecordOnSwitch
    record_of_output_node: RecordOnSwitch

File: univ/activeMeasurement/AuxiliaryCode/数据结构.cpp
#include <map>
#include <vector>

typedef uint32_t TimeStamp;
typedef uint32_t LinkID;
typedef uint32_t SwitchID;
typedef uint32_t FlowID;
typedef uint32_t IPID;

class RecordPerSwitch{
    IPID IP_id;
    TimeStamp time_stamp;
};

class PathRecord{
    // 长度和对应的Flow.path相同
    std::vector<RecordPerSwitch> record;
};

class Flow{
    FlowID flow_id;
    // 标识每一跳交换机的id
    std::vector<SwitchID> path;
    // 标识每次路径的信息
    std::vector<PathRecord> records;
};

class RecordOnSwitch{
    std::map<FlowID, std::map<IPID, TimeStamp> > record;
};

class Link{
    LinkID link_id;
    RecordOnSwitch record_of_input_node;
    RecordOnSwitch record_of_output_node;
};